test/datasets_utils.py

```diff
@@ -0,0 +1,614 @@
+                pass
+
+    def test_smoke(self):
+        with self.create_dataset() as (dataset, _):
+            self.assertIsInstance(dataset, torchvision.datasets.VisionDataset)
+
+    @test_all_configs
+    def test_str_smoke(self, config):
+        with self.create_dataset(config) as (dataset, _):
+            self.assertIsInstance(str(dataset), str)
+
+    @test_all_configs
+    def test_feature_types(self, config):
+        with self.create_dataset(config) as (dataset, _):
+            example = dataset[0]
+
+            actual = len(example)
+            expected = len(self.FEATURE_TYPES)
+            self.assertEqual(
+                actual,
+                expected,
+                f"The number of the returned features does not match the the number of elements in in FEATURE_TYPES: "
+                f"{actual} != {expected}",
+            )
+
+            for idx, (feature, expected_feature_type) in enumerate(zip(example, self.FEATURE_TYPES)):
+                with self.subTest(idx=idx):
+                    self.assertIsInstance(feature, expected_feature_type)
+
+    @test_all_configs
+    def test_num_examples(self, config):
+        with self.create_dataset(config) as (dataset, info):
+            self.assertEqual(len(dataset), info["num_examples"])
+
+
+class ImageDatasetTestCase(DatasetTestCase):
+    """Abstract base class for image dataset testcases.
+
+    - Overwrites the FEATURE_TYPES class attribute to expect a :class:`PIL.Image.Image` and an integer label.
+    """
+
+    FEATURE_TYPES = (PIL.Image.Image, int)
+
+    @contextlib.contextmanager
+    def create_dataset(
+        self,
+        config: Optional[Dict[str, Any]] = None,
+        inject_fake_data: bool = True,
+        disable_download_extract: Optional[bool] = None,
+        **kwargs: Any,
+    ) -> Iterator[Tuple[torchvision.datasets.VisionDataset, Dict[str, Any]]]:
+        with super().create_dataset(
+            config=config,
+            inject_fake_data=inject_fake_data,
+            disable_download_extract=disable_download_extract,
+            **kwargs,
+        ) as (dataset, info):
+            # PIL.Image.open() only loads the image meta data upfront and keeps the file open until the first access
+            # to the pixel data occurs. Trying to delete such a file results in an PermissionError on Windows. Thus, we
+            # track all lazily opened images and close the file handle before the file is deleted.
+            # This problem only occurs during testing since some tests, e.g. DatasetTestCase.test_feature_types open an
+            # image, but never use the underlying data. During normal operation it is reasonable to assume that the
+            # user wants to work with the image he just opened rather than deleting the underlying file.
+            with self._close_image_handles():
+                yield dataset, info
+
+    @contextlib.contextmanager
+    def _close_image_handles(self):
```

Comment:
This is a bit convoluted, but I assume doing `PIL.Image.open().convert('RGB')` (or something equivalent that reads the data) is not enough to fix Windows issues?

Reply:
I agree. I actually wanted monkeypatch the `__getitem__` method to simply `.load()` all `PIL.Image.Image`s, but unfortunately this is [not possible](https://stackoverflow.com/questions/28355896/monkeypatch-with-instance-method). Thus, I reverted to record all calls to `PIL.Image.open()` and close the file handles before the file is deleted. I'm open to ideas though.

Reply:
After an internal discussion, we agreed on patching `PIL.Image.open()` globally. Thus, there is no need to discover the `PIL.Image` module in the corresponding dataset module. Furthermore, instead of tracking open files and closing them, we simply `.load()` them, since the amount of data / images is fairly low. Combined this should make it more readable.